inc/knowbaseitem.class.php

```diff
@@ -399,99 +399,55 @@ public function haveVisibilityAccess() {
    * Return visibility joins to add to SQL
    *
    * @since 0.83
+   * @deprecated 9.4.0
    *
    * @param $forceall force all joins (false by default)
    *
    * @return string joins to add
    **/
    static function addVisibilityJoins($forceall = false) {
+      global $DB;
 
-      $join = '';
-
-      // Users
-      $join .= " LEFT JOIN `glpi_knowbaseitems_users`
-                     ON (`glpi_knowbaseitems_users`.`knowbaseitems_id` = `glpi_knowbaseitems`.`id`) ";
+      Toolbox::deprecated('Use getVisibilityCriteria');
 
-      // Groups
-      if ($forceall
-          || (isset($_SESSION["glpigroups"]) && count($_SESSION["glpigroups"]))) {
-         $join .= " LEFT JOIN `glpi_groups_knowbaseitems`
-                        ON (`glpi_groups_knowbaseitems`.`knowbaseitems_id`
-                              = `glpi_knowbaseitems`.`id`) ";
-      }
+      //get and clean criteria
+      $criteria = self::getVisibilityCriteria();
+      unset($criteria['WHERE']);
+      $criteria['FROM'] = self::getTable();
 
-      // Profiles
-      if ($forceall
-          || (isset($_SESSION["glpiactiveprofile"])
-              && isset($_SESSION["glpiactiveprofile"]['id']))) {
-         $join .= " LEFT JOIN `glpi_knowbaseitems_profiles`
-                        ON (`glpi_knowbaseitems_profiles`.`knowbaseitems_id`
-                              = `glpi_knowbaseitems`.`id`) ";
-      }
-
-      // Entities
-      if ($forceall
-          || !Session::getLoginUserID()
-          || (isset($_SESSION["glpiactiveentities"]) && count($_SESSION["glpiactiveentities"]))) {
-         $join .= " LEFT JOIN `glpi_entities_knowbaseitems`
-                        ON (`glpi_entities_knowbaseitems`.`knowbaseitems_id`
-                              = `glpi_knowbaseitems`.`id`) ";
-      }
-
-      return $join;
+      $it = new \DBmysqlIterator(null);
+      $it->execute($criteria);
```

Comment:
This query should just be evaluated, not executed.
```suggestion
      $it->buildQuery($criteria);
```